ledger/simulation/resources.go

```diff
@@ -605,3 +607,696 @@ func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operatio
+
+		p.txnResources[groupIndex].prefilledAccounts = append(p.txnResources[groupIndex].prefilledAccounts, txn.Accounts...)
+
+		for _, box := range txn.Boxes {
+			ref := logic.BoxRef{App: txn.ForeignApps[box.Index], Name: string(box.Name)}
+			p.txnResources[groupIndex].prefilledBoxes = append(p.txnResources[groupIndex].prefilledBoxes, ref)
+		}
+
+		p.txnResources[groupIndex].appFromField = txn.ApplicationID
+
+	case protocol.AssetTransferTx:
+		p.txnResources[groupIndex].assetFromField = txn.XferAsset
+		p.txnResources[groupIndex].addAddressFromField(txn.AssetReceiver)
+		p.txnResources[groupIndex].addAddressFromField(txn.AssetCloseTo)
+		p.txnResources[groupIndex].addAddressFromField(txn.AssetSender)
+
+	case protocol.PaymentTx:
+		p.txnResources[groupIndex].addAddressFromField(txn.Receiver)
+		p.txnResources[groupIndex].addAddressFromField(txn.CloseRemainderTo)
+
+	case protocol.AssetConfigTx:
+		p.txnResources[groupIndex].assetFromField = txn.ConfigAsset
+
+	case protocol.AssetFreezeTx:
+		p.txnResources[groupIndex].assetFromField = txn.FreezeAsset
+		p.txnResources[groupIndex].addAddressFromField(txn.FreezeAccount)
+	}
+}
+
+func (p *resourcePopulator) addAccount(addr basics.Address) error {
+	var err error
+
+	// If another txn has the account, do nothing
+	// This should never happen because the logic in EvalContext should
+	// prevent duplicate resources in the tracker, but we check here just to be safe
+	for _, txn := range p.txnResources {
+		if txn.hasAccount(addr) {
+			return nil
+		}
+	}
+
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addAccount(addr)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) addAsset(asset basics.AssetIndex) error {
+	var err error
+
+	// If another txn has the asset, do nothing
+	// This should never happen because the logic in EvalContext should
+	// prevent duplicate resources in the tracker, but we check here just to be safe
+	for _, txn := range p.txnResources {
+		if txn.hasAsset(asset) {
+			return nil
+		}
+	}
+
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addAsset(asset)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) addApp(app basics.AppIndex) error {
+	var err error
+
+	// If another txn has the app, do nothing
+	// This should never happen because the logic in EvalContext should
+	// prevent duplicate resources in the tracker, but we check here just to be safe
+	for _, txn := range p.txnResources {
+		if txn.hasApp(app) {
+			return nil
+		}
+	}
+
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addApp(app)
+		if err == nil {
+			return nil
+		}
+	}
+	return err
+}
+
+func (p *resourcePopulator) addBox(app basics.AppIndex, name string) error {
+	var err error
+
+	// First try to find txn with app already available
+	for _, i := range p.appCallIndexes {
+		if app == basics.AppIndex(0) || p.txnResources[i].hasApp(app) {
```

Comment:
This seems strange, can you add a comment explaining how you are handling AppIndex(0)?  When does `addBox(0, ...)` get called?  I would guess that if this is getting called, it's because a specific app call, maybe a create, needs it. So you can't check all the appCallIndexes for `(0,...)` because if you find it elsewhere, it means something else.